### ISSUE TITLE
Signature of Array#slice! is Incorrect

### ISSUE THREAD

**Issue by toddsundsted** [sorbet/sorbet#1988]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%5B%5D.slice!(0%2C%20-1)%20%7C%7C%20%5B%5D%0A)

```ruby
# typed: true
[].slice!(0, -1) || []
```

#### Observed output

```
editor.rb:2: This code is unreachable https://srb.help/7006
     2 |[].slice!(0, -1) || []
                            ^^
Errors: 1
```

#### Expected behavior

Calls to `slice!` may return `nil`.

```
2.6.3 :001 > [].slice!(0, -1)
 => nil
2.6.3 :002 > [].slice!(0, -1) || []
 => []
```

[From the documentation](https://ruby-doc.org/core-2.6.5/Array.html#method-i-slice-21):

```
slice!(index) → obj or nil
slice!(start, length) → new_ary or nil
slice!(range) → new_ary or nil 
```

- - -

The changes to core/array.rbi should be as simple as:

```
  sig do
    params(
        arg0: T::Range[Integer],
    )
    .returns(T.nilable(T::Array[Elem]))
  end
  sig do
    params(
        arg0: Integer,
        arg1: Integer,
    )
    .returns(T.nilable(T::Array[Elem]))
  end
  sig do
    params(
        arg0: T.any(Integer, Float),
    )
    .returns(T.nilable(Elem))
  end
  def slice!(arg0, arg1=T.unsafe(nil)); end
```

(I haven't managed to get Sorbet to build, so I can't verify this change myself yet.)

**Comment by toddsundsted** [comment]
Fix proposed in https://github.com/sorbet/sorbet/pull/1989

**Comment by jez** [comment]
Fixed in #1989
